File: Frontend/highlight.py
from dataclasses import dataclass
from typing import Optional
from clickTypeEnum import ClickType
from constants import SPAWN_CODE
from default_abilities import VISUALS
from drawClasses import IDItem


@dataclass
class Highlight:
    item: Optional[IDItem] = None
    usage: Optional[int] = None

    def wipe(self):
        self.item = None
        self.usage = None

    @property
    def color(self):
        if self.usage:
            if self.usage in set(VISUALS) - {SPAWN_CODE}:
                return VISUALS[self.usage].color
        return None
    
    @property
    def type(self):
        if self.item:
            return self.item.type
        return ClickType.BLANK

    def send_format(self, items=None, code=None):
        code = code or self.usage
        if not items:
            items = [self.item.id]
        else:
            print(items)
        return {"code": code, "items": items}
    
    def __call__(self, new_item: IDItem, new_usage: int):
        self.item = new_item
        self.usage = new_usage

    def __bool__(self):
        return bool(self.item)

File: TypeScriptFrontend/highlight.ts
import { IDItem } from "./idItem"; // Assume these are defined in the conversation// Assumed imports based on prior context
import { ClickType } from "./enums"; // Assume this is defined
import { AbilityVisual } from "./immutable_visuals"; // Assume this is defined
import { VISUALS } from "./default_abilities"; // Assume this is defined
import { Colors, KeyCodes } from "./constants";

export class Highlight {
    item: IDItem | null = null;
    usage: number | null = null;

    wipe(): void {
        this.item = null;
        this.usage = null;
    }

    get color(): readonly [number, number, number] {
        if (this.usage && this.usage !== KeyCodes.SPAWN_CODE) {
            const visual = VISUALS[this.usage] as AbilityVisual; // Assuming AbilityVisual has a color property
            return visual.color;
        }
        return Colors.GREY;
    }

    get type(): ClickType {
        if (this.item) {
            return this.item.type;
        }
        return ClickType.BLANK;
    }

    sendFormat(items?: number[], code?: number): object {
        const coda = code ?? this.usage;
        items = items || (this.item ? [this.item.id] : []);
        return { code, items };
    }

    valueOf(): boolean {
        return Boolean(this.item);
    }
}